lib/reek/smells/instance_variable_assumption.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+require_relative 'smell_detector'
+require_relative 'smell_warning'
+
+module Reek
+  module Smells
+    #
+    # The +InstanceVariableAssumption+ class is responsible for
+    # detecting directly access of instance variables in a class
+    # that does not define them in its initialize method.
+    #
+    class InstanceVariableAssumption < SmellDetector
+      def self.contexts
+        [:class]
+      end
+
+      # Checks +klass+ for instance
+      # variables assumptions.
+      #
+      # @return [Array<SmellWarning>]
+      #
+      def inspect(ctx)
+        ivar_assumptions = class_ivars(ctx) - initialized_ivars(ctx)
+
+        return [] if ivar_assumptions.empty?
+
+        build_smell_warning(ctx, ivar_assumptions.uniq)
+      end
+
+      private
+
+      def build_smell_warning(ctx, ivar_assumptions)
+        [smell_warning(
+          context: ctx,
+          lines: [ctx.exp.line],
+          message: "has #{ivar_assumptions.join(', ')} instance variable assumptions",
+          parameters: { assumptions: ivar_assumptions })]
+      end
+
+      # Collection of ivars used by class
+      # :reek:UtilityFunction
+      def class_ivars(ctx)
+        ctx.node_instance_methods.map(&:instance_variable_assignments).flatten
+      end
+
+      # Collection of ivars found on initialize
```

Comment:
Same as below. Method documentation is missing. "ctx" doesn't have to be explained (otherwise you'd have to do it in all smell detectors and it would be redundant) but we need to be crystal clear about the return types.